# Does[0]Compute Fall 2016 &ndash; Spring 2017 Syllabus

## Topics Covered

 - File Conversions and Other Utilities
 - Sequence Cleaning and Quality Control
 - Read Mapping with Short and Long Read Aligners
 - SAM Processing
 - Genomic Ranges
 - Variant Detection
 - Variant Annotation
 - Quantitiave Genetics
 - Population Genetics

## Tentative Dates
We meet every Tuesday from 2:00 PM to 3:00 PM in 213 [Boralug Hall](http://campusmaps.umn.edu/borlaug-hall) for the Spring 2017 semester. Starting May 16<sup>th</sup>, we meet every other Tuesday from 2:00 PM to 3:00 PM in 213 [Boralug Hall](http://campusmaps.umn.edu/borlaug-hall).

---

## Summer 2017 Meetings and Materials

### May 16<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) various frameworks
#### Group Discussion: Jigsaw Learning Strategy
Each person will pick one paper to read before the discussion. Groups will be divided based on the paper chosen.
#### Materials:
- [Thornton (2014). A C++ Template Library for Efficient Forward-Time Population Genetic Simulation of Large Populations. *Genetics*.](http://www.genetics.org/content/198/1/157)
- [`fwdpy` Documentation](https://molpopgen.github.io/fwdpy_manual/index.html)
- [Thornton et al. (2013). Properties and Modeling of GWAS when Complex Disease Risk is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. *PLOS Genetics*.](http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003258)
- [Baldwin-Brown et al. (2014). The Power to Detect Quantitative Trait Loci Using Resequenced, Experimentally Evolved Populations of Diploid, Sexual Organisms. *Molecular Biology and Evolution*.](https://academic.oup.com/mbe/article/31/4/1040/1108505/The-Power-to-Detect-Quantitative-Trait-Loci-Using)

### May 30<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) simulation
#### Discussion Leader: TBD
#### Materials: TBD

### June 13<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) simulation
#### Discussion Leader: TBD
#### Materials: TBD

### June 27<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) simulation
#### Discussion Leader: TBD
#### Materials: TBD

### July 11<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) project
#### Discussion Leader: TBD
#### Materials: TBD

### July 25<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) project
#### Discussion Leader: TBD
#### Materials: TBD

### August 8<sup>th</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) project
#### Discussion Leader: TBD
#### Materials: TBD

### August 22<sup>nd</sup>: [`fwdpp`](http://molpopgen.github.io/fwdpp/)/[`fwdpy`](http://molpopgen.github.io/fwdpy/)/[`fwdpy11`](https://molpopgen.github.io/fwdpy11/) project
#### Discussion Leader: TBD
#### Materials: TBD

---

## Spring 2017 Meetings and Materials

### January 24<sup>th</sup>, 2017: [`sequence_handling`](https://github.com/MorrellLAB/sequence_handling)
#### Discussion Leader: Fernanda Rodriguez

### January 31<sup>st</sup>: CANCELED

### February 7<sup>st</sup>: Variant Detection
#### Discussion Leader: Li Lei
 - [GATK](https://software.broadinstitute.org/gatk/) Part I
   - [Paper](http://www.ncbi.nlm.nih.gov/pubmed/20644199)

### February 14<sup>th</sup>: Variant Detection
#### Discussion Leader: Peter Morrell
 - [Freebayes](https://github.com/ekg/freebayes)
   - [Paper](https://arxiv.org/pdf/1207.3907v2.pdf)
 - [SAMTools mpileup](http://www.htslib.org/doc/samtools.html)/[bcftools](http://www.htslib.org/doc/bcftools.html)
   - [SAMTools SNP Calling Paper](http://www.ncbi.nlm.nih.gov/pubmed/21903627)
   - [bcftools Paper](http://samtools.github.io/bcftools/call-m.pdf)
   - [SAMTools Mapping to Variant Calls](http://www.htslib.org/workflow/#mapping_to_variant)

### February 21<sup>st</sup>: Variant Detection
#### Discussion Leader: Paul Hoffman
 - [VCFtools](https://vcftools.github.io/index.html)
   - [Paper](http://bioinformatics.oxfordjournals.org/content/27/15/2156)
 - [Platypus](http://www.well.ox.ac.uk/platypus)
   - [Paper](http://www.nature.com/ng/journal/v46/n8/full/ng.3036.html)
 - [Summary Paper](http://www.nature.com/articles/srep17875)

### February 28<sup>th</sup>: Variant Annotation
#### Discussion Leader: Tom Kono
 - [BAD_Mutations](https://github.com/MorrellLAB/BAD_Mutations)
 - [ANNOVAR](http://annovar.openbioinformatics.org/en/latest/)
   - [Paper](http://nar.oxfordjournals.org/content/38/16/e164)

### March 7<sup>th</sup>: Quantitiave Genetics
#### Discussion Leader: Li Lei
 - [R/QTL](http://rqtl.org) Part I
   - [Paper](http://bioinformatics.oxfordjournals.org)

### March 14<sup>th</sup>: Quantitiave Genetics
#### Discussion Leader: Tom Kono
 - [R/QTL](http://rqtl.org) Part II
   - [Paper](http://bioinformatics.oxfordjournals.org/content/19/7/889.long)

### March 21<sup>st</sup>: Quantitiave Genetics
#### Discussion Leader: Alex Olhoff
 - [R NAM](https://cran.r-project.org/web/packages/NAM/index.html)

### March 28<sup>th</sup>: Big Data Visualization
#### Discussion Leader: Alex Olhoff

### April 4<sup>th</sup>:

Break, we will resume weekly discussions in two weeks on April 18th.

### April 11<sup>th</sup>:

Break, we will resume weekly discussions next week on April 18th.

### April 18<sup>th</sup>: Population Genetics
#### Discussion Leader: Chaochih Liu
 - [ANGSD](http://www.popgen.dk/wiki/index.php/ANGSD)/[ANGSD-wrapper](https://github.com/mojaveazure/angsd-wrapper) Part I
   - [ANGSD Paper](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-014-0356-4)
   - [ANGSD-wrapper Paper](http://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12578/abstract;jsessionid=A7491A25E5383E19EEC7193A9D3AF8E4.f01t03)

### April 25<sup>th</sup>: Population Genetics
#### Discussion Leader: Chaochih Liu
 - [ANGSD](http://www.popgen.dk/wiki/index.php/ANGSD)/[ANGSD-wrapper](https://github.com/mojaveazure/angsd-wrapper) Part II
   - [ANGSD Paper](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-014-0356-4)
   - [ANGSD-wrapper Paper](http://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12578/abstract;jsessionid=A7491A25E5383E19EEC7193A9D3AF8E4.f01t03)

### May 2<sup>nd</sup>: 10xgenomics data analysis
#### Discussion Leader:John R Garbe in UMGC
 - [slides](https://github.com/MorrellLAB/DoesNaughtCompute/tree/master/10xgenomic_sequencing)

### May 9<sup>th</sup>: Quantitiave Genetics
#### Discussion Leader: Peter Morrell
 - [BEAGLE](http://faculty.washington.edu/browning/beagle/beagle.html)
   - [Phasing Paper](http://www.sciencedirect.com/science/article/pii/S0002929707638828)
   - [Imputing Paper](http://www.sciencedirect.com/science/article/pii/S0002929715004917)
   - [IBD Detection Paper](http://www.genetics.org/content/194/2/459)
 - [PHASE](http://stephenslab.uchicago.edu/software.html#phase)/[fastPHASE](http://stephenslab.uchicago.edu/software.html#fastphase)
   - [PHASE Paper](http://stephenslab.uchicago.edu/assets/papers/Stephens2003a.pdf)
   - [fastPHASE Paper](http://stephenslab.uchicago.edu/assets/papers/Scheet2006.pdf)
 - other reference papers
   - [Allele-specific PCR can improve the efficiency of experimental resolution of heterozygotes in resequencing studies](http://onlinelibrary.wiley.com/doi/10.1111/j.1755-0998.2009.02804.x/abstract)
   - [Error detection in SNP data by considering the likelihood of recombinational history implied by three-site combinations](https://academic.oup.com/bioinformatics/article/23/14/1807/190632/Error-detection-in-SNP-data-by-considering-the)

### May 16<sup>th</sup>: Population Genetics
#### Discussion Leader: TBD
 - [hierfstat](https://cran.r-project.org/web/packages/hierfstat/index.html)
   - [Paper](http://www2.unil.ch/popgen/research/reprints/goudet_men_2005.pdf)
 - [LDhat](https://github.com/auton1/LDhat)

### May 23<sup>rd</sup>: Population Genetics
#### Discussion Leader: TBD
 - [STRUCTURE](http://pritchardlab.stanford.edu/structure.html)
   - [Paper](http://pritchardlab.stanford.edu/publications/pdfs/PritchardEtAl00.pdf)
 - [ADMIXTURE](https://www.genetics.ucla.edu/software/admixture/)
   - [Paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2752134/)

---

## Fall 2016 Meetings and Materials
### September 22<sup>nd</sup>, 2016: File Conversions and Other Utilities
#### Discussion Leader: Li Lei and Paul Hoffman
 - [UNIX `sort`](http://www.wikiwand.com/en/Sort_(Unix))
 - [Plink 1.9](https://www.cog-genomics.org/plink2) Part I
   - [Paper](http://gigascience.biomedcentral.com/articles/10.1186/s13742-015-0047-8)

### September 29<sup>th</sup>, 2016: File Conversions and Other Utilities
#### Discussion Leader: Chaochih Liu
 - [Plink 1.9](https://www.cog-genomics.org/plink2) Part II
   - [Paper](http://gigascience.biomedcentral.com/articles/10.1186/s13742-015-0047-8)

### October 6<sup>th</sup>, 2016: Sequence Cleaning and Quality Control
#### Discussion Leader: Paul Hoffman
 - [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
 - [FASTX-toolkit](http://hannonlab.cshl.edu/fastx_toolkit/)

### October 13<sup>th</sup>, 2016: Sequence Cleaning and Quality Control
#### Discussion Leader: Chaochih Liu
 - [Seqqs](https://github.com/vsbuffalo/seqqs)
 - [Sickle](https://github.com/najoshi/sickle)
 - [Scythe](https://github.com/vsbuffalo/scythe)

### October 20<sup>th</sup>, 2016: Short Read Aligners
#### Discussion Leader: Peter Morrell
 - **Recommended Reading**
   - Clegg and Morrell, 2004, [Bioinformatics](https://github.com/MorrellLAB/DoesNaughtCompute/raw/master/Short_Read_Mapping_Part_I/Clegg-_-Morrell-04-_bioinformatics_.pdf)
   - Clegg and Morrell, 2004, [Mutational Processes](https://github.com/MorrellLAB/DoesNaughtCompute/raw/master/Short_Read_Mapping_Part_I/Clegg-_-Morrell-04-_mutation_.pdf)
 - [BWA](http://bio-bwa.sourceforge.net)
 - [Bowtie 2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
   - [Paper](http://www.nature.com/nmeth/index.html)

### October 27<sup>th</sup>, 2016: Short Read Aligners
#### Discussion Leader: Ying Zhang
 - [Benchmarking short sequence mapping tools](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-184)
 - [BWA](http://bio-bwa.sourceforge.net)
 - [Bowtie 2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
   - [Paper](http://www.nature.com/nmeth/journal/v9/n4/full/nmeth.1923.html)


### Novermber 3<sup>rd</sup>, 2016: Long Read Aligners
#### Discussion Leader: Kevin Silverstein and Corey Carter
 - [BLASR](https://github.com/PacificBiosciences/blasr)
 - [MHAP](http://mhap.readthedocs.io/en/latest/index.html)
 - [DALINGER](https://github.com/thegenemyers/DALIGNER)
   - [Comparison Paper](http://www.nature.com/nbt/journal/v33/n6/full/nbt.3238.html)

### November 10<sup>th</sup>, 2016: SAM Processing
#### Discussion Leader: Li Lei
 - [SAMTools](http://www.htslib.org/doc/samtools.html)
   - [Paper](http://www.ncbi.nlm.nih.gov/pubmed/19505943)
 - [Picard](http://broadinstitute.github.io/picard/)

### November 17<sup>th</sup>, 2016: SAM Processing
#### Discussion Leader: Colin Pierce
 - [SAMTools](http://www.htslib.org/doc/samtools.html)
   - [Paper](http://www.ncbi.nlm.nih.gov/pubmed/19505943)
 - [Picard](http://broadinstitute.github.io/picard/)


### December 1<sup>st</sup>, 2016: Advanced Software Installation
#### Discussion Leader: Paul Hoffman
 - [GCC/G++](https://gcc.gnu.org/)
 - [`configure`](https://www.wikiwand.com/en/Configure_script)
 - [GNU Make](http://www.gnu.org/software/make/)
 - [CMake](https://cmake.org/)
 - Java programs
 - [`setup.py`](https://docs.python.org/3.5/distutils/setupscript.html)
 - [Modulefiles](http://modules.sourceforge.net/)

### December 8<sup>th</sup>, 2016: **CANCELED**

### December 15<sup>th</sup>, 2016: Genomic Ranges
#### Discussion Leader: Chaochih Liu
 - [BEDTools](http://bedtools.readthedocs.io/en/latest/)
   - [Paper](http://bioinformatics.oxfordjournals.org/content/26/6/841.short)
 - [BEDOPS](https://bedops.readthedocs.io/en/latest/)
   - [Paper](http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract)
